apps/ledger-live-mobile/src/screens/FirmwareUpdate/useUpdateFirmwareAndRestoreSettings.ts

```diff
@@ -0,0 +1,113 @@
+import staxLoadImage from "@ledgerhq/live-common/hw/staxLoadImage";
+import staxFetchImage from "@ledgerhq/live-common/hw/staxFetchImage";
+import { createAction as createStaxLoadImageAction } from "@ledgerhq/live-common/hw/actions/staxLoadImage";
+import { createAction as createStaxFetchImageAction } from "@ledgerhq/live-common/hw/actions/staxFetchImage";
+import { useUpdateFirmware } from "@ledgerhq/live-common/deviceSDK/hooks/useUpdateFirmware";
+import { Device, DeviceModelId } from "@ledgerhq/types-devices";
+import {
+  UpdateFirmwareActionState,
+  updateFirmwareActionArgs,
+} from "@ledgerhq/live-common/deviceSDK/actions/updateFirmware";
+import { Observable } from "rxjs";
+import { useEffect, useState } from "react";
+
+export type FirmwareUpdateParams = {
+  device: Device;
+  updateFirmwareAction?: (
+    args: updateFirmwareActionArgs,
+  ) => Observable<UpdateFirmwareActionState>;
+};
+
+export type UpdateStep =
+  | "start"
+  | "imageBackup"
+  | "firmwareUpdate"
+  | "languageRestore"
+  | "imageRestore"
+  | "appsRestore"
+  | "completed";
+
+const staxLoadImageAction = createStaxLoadImageAction(staxLoadImage);
+const staxFetchImageAction = createStaxFetchImageAction(staxFetchImage);
+
+export const useUpdateFirmwareAndRestoreSettings = ({
+  updateFirmwareAction,
+  device,
+}: FirmwareUpdateParams) => {
+  const [updateStep, setUpdateStep] = useState<UpdateStep>("start");
+
+  const staxFetchImageState = staxFetchImageAction.useHook(
+    updateStep === "imageBackup" ? device : null,
+    "",
+  );
+
+  const { triggerUpdate, updateState: updateActionState } = useUpdateFirmware({
+    deviceId: device?.deviceId ?? "",
+    updateFirmwareAction,
+  });
+
+  const staxLoadImageState = staxLoadImageAction.useHook(
+    updateStep === "imageRestore" && staxFetchImageState.hexImage
+      ? device
+      : null,
+    staxFetchImageState.hexImage ?? "",
+    false
+  );
+
+  useEffect(() => {
+    switch (updateStep) {
+      case "start":
+        if (device.modelId === DeviceModelId.stax) {
+          setUpdateStep("imageBackup");
+        } else {
+          setUpdateStep("firmwareUpdate");
+        }
+        break;
+      case "imageBackup":
+        if (staxFetchImageState.imageFetched || staxFetchImageState.error) {
+          // TODO: check if we want to do something with error, maybe just log it
+          setUpdateStep("firmwareUpdate");
+        }
+        break;
+      case "firmwareUpdate":
+        if (updateActionState.step === "preparingUpdate") {
+          triggerUpdate();
+        } else if (updateActionState.step === "firmwareUpdateCompleted") {
+          setUpdateStep("languageRestore");
+        }
+        break;
+      case "languageRestore":
+        // TODO: implement image restore
+        setUpdateStep("imageRestore");
+        break;
```

Comment:
case and step dont match? or is it like _next_ step kind of deal, confusing

Reply:
It's a "next" step kind of deal. This useEffect implements a state machine. I can add some comments explaining that

Reply:
I've refactored it in https://github.com/LedgerHQ/ledger-live/pull/3264 there I separated these changes into functions with a clear name of their purpose. Is it ok for you if we leave this one like this for now?